client/src/main/scala/org/http4s/client/PoolManager.scala

```diff
@@ -84,30 +120,36 @@ private final class PoolManager[F[_], A <: Connection[F]](
         if (!isClosed) {
           @tailrec
           def go(): Unit =
-            idleQueue.dequeueFirst(_.requestKey == key) match {
+            getConnectionFromQueue(key) match {
               case Some(conn) if !conn.isClosed =>
                 logger.debug(s"Recycling connection: $stats")
                 callback(Right(NextConnection(conn, fresh = false)))
 
               case Some(closedConn) =>
                 logger.debug(s"Evicting closed connection: $stats")
-                allocated -= 1
+                decrConnection(key)
                 go()
 
-              case None if allocated < maxTotal =>
+              case None if numConnectionsCheckHolds(key) =>
                 logger.debug(s"Active connection not found. Creating new one. $stats")
                 createConnection(key, callback)
 
-              case None if idleQueue.nonEmpty =>
+              case None if curTotal == maxTotal =>
                 logger.debug(
                   s"No connections available for the desired key. Evicting oldest and creating a new connection: $stats")
-                allocated -= 1
-                idleQueue.dequeue().shutdown()
-                createConnection(key, callback)
+                val keys = idleQueues.keys
+                if (keys.nonEmpty) {
+                  val randKey = keys.drop(Random.nextInt(keys.size)).head
```

Comment:
How about using keys the iterator so we don't create an intermediate collection:
```scala
val randKey = keys.iterator.drop(Random.nextInt(keys.size)).next()
...
```

Reply:
Where is the intermediate collection?

Reply:
It's made by the `drop` call on `keys`.

Reply:
Ok